dracut/modules.d/99asahi-firmware/install-asahi-firmware.sh

```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+# SPDX-License-Identifier: MIT
+
+type getarg > /dev/null 2>&1 || . /lib/dracut-lib.sh
+
+info ":: Asahi: Updating vendor firmware in root filesystem..."
+mkdir -p "${NEWROOT}/lib/firmware"
+TARGET_ROOT="${NEWROOT}" update-vendor-firmware
```

Comment:
This doesn't work properly because the root filesystem is read only at this stage

Reply:
I've refreshed this to use the `cleanup` stage. Could you try to see if that works properly?

Reply:
Nope, same failure (cannot open read only filesystem).

Reply:
We might need to do remounts then... Any suggestions on logic for this?

Reply:
@johannbg, @haraldh, @lnykryn: Any ideas?

Reply:
> We might need to do remounts then... Any suggestions on logic for this?
I've experimented with this a bunch and I'm not convinced it can be safely done in the initrd environment. Among other things, it could trigger a fsck if the filesystem is unclean. What we probably need to do is to move this to a systemd unit ordered right after the "remount / as rw" action happens. @daandemeyer, any suggestions for what we could use here?

Reply:
Ok, I figured this out. We drop `install-vendor-firmware.sh` altogether, have `load-vendor-firmware.sh` cache the firmware in `/run`, and then add a systemd unit to load it from there, so we don't need to mess with the ESP again:

```diff
diff --git a/dracut/modules.d/99asahi-firmware/install-asahi-firmware.sh b/dracut/modules.d/99asahi-firmware/install-asahi-firmware.sh
deleted file mode 100755
index f163a37..0000000
--- a/dracut/modules.d/99asahi-firmware/install-asahi-firmware.sh
+++ /dev/null
@@ -1,8 +0,0 @@
-#!/bin/sh
-# SPDX-License-Identifier: MIT
-
-type getarg > /dev/null 2>&1 || . /lib/dracut-lib.sh
-
-info ":: Asahi: Updating vendor firmware in root filesystem..."
-mkdir -p "${NEWROOT}/lib/firmware"
-TARGET_ROOT="${NEWROOT}" update-vendor-firmware
diff --git a/dracut/modules.d/99asahi-firmware/load-asahi-firmware.sh b/dracut/modules.d/99asahi-firmware/load-asahi-firmware.sh
index 1b7f864..bd5fc09 100755
--- a/dracut/modules.d/99asahi-firmware/load-asahi-firmware.sh
+++ b/dracut/modules.d/99asahi-firmware/load-asahi-firmware.sh
@@ -25,5 +25,4 @@ fi
 # If the above exists, hopefully the /dev device exists and this will work
 info ":: Asahi: Unpacking vendor firmware into initramfs..."
 mkdir -p /lib/firmware
-update-vendor-firmware
-
+CACHE=true update-vendor-firmware
diff --git a/dracut/modules.d/99asahi-firmware/module-setup.sh b/dracut/modules.d/99asahi-firmware/module-setup.sh
index 3c4f0f4..db63ca0 100755
--- a/dracut/modules.d/99asahi-firmware/module-setup.sh
+++ b/dracut/modules.d/99asahi-firmware/module-setup.sh
@@ -30,5 +30,4 @@ install() {
     $DRACUT_CP -R -L -t "${initdir}/${asahiscriptsdir}" "${dracutsysrootdir}${asahiscriptsdir}"/*
     inst_multiple comm cut dirname modprobe seq sha1sum sleep tar update-vendor-firmware
     inst_hook pre-udev 10 "${moddir}/load-asahi-firmware.sh"
-    inst_hook cleanup 90 "${moddir}/install-asahi-firmware.sh"
 }
diff --git a/systemd/update-vendor-firmware.service b/systemd/update-vendor-firmware.service
new file mode 100644
index 0000000..8a9aa8b
--- /dev/null
+++ b/systemd/update-vendor-firmware.service
@@ -0,0 +1,17 @@
+[Unit]
+Description=Update vendor firmware
+DefaultDependencies=no
+Conflicts=shutdown.target
+After=systemd-remount-fs.target
+Before=systemd-udev-trigger.service sysinit.target shutdown.target
+ConditionDirectoryNotEmpty=/run/vendorfw
+
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+Environment=VENDORFW=/run/vendorfw
+ExecStart=update-vendor-firmware
+ExecStart=rm -r /run/vendorfw
+
+[Install]
+WantedBy=sysinit.target
diff --git a/update-vendor-firmware b/update-vendor-firmware
index 541d947..1f8088c 100755
--- a/update-vendor-firmware
+++ b/update-vendor-firmware
@@ -17,6 +17,7 @@ fi
 : ${TARGET_ROOT:=/}
 : ${TARGET:=/lib/firmware/}
 : ${TARGET_MANIFEST:=".vendorfw.manifest"}
+: ${CACHE:=}
 
 umount=false
 
@@ -56,8 +57,14 @@ if [ -e "$TARGET_MANIFEST" ]; then
     fi
 fi
 
+if [ -n "$CACHE" ]; then
+  echo "Caching vendor firmware..."
+  cp -pr "$VENDORFW" /run
+  VENDORFW=/run/vendorfw
+fi
+
 echo "Extracting updated vendor firmware..."
-tar xf "$VENDORFW/firmware.tar"
+tar -x --overwrite -f "$VENDORFW/firmware.tar"
 
 if [ -e "$TARGET_MANIFEST" ]; then
     echo "Cleaning up obsolete firmware..."
```

Reply:
Note that diff also includes https://github.com/AsahiLinux/asahi-scripts/pull/10 and is missing the makefile integration (this systemd unit should be installed only for Fedora).

Reply:
Updated #10 with the systemd and update-vendor-firmware changes needed.

Reply:
I've revised this entirely based on yours and @chadmed's suggestions.